### DISCUSSION TITLE
Whitelist helper function works for /etc but not for the home directory as mentioned in the wiki

### DISCUSSION THREAD

**Discussion by saipavanc** [CyberShadow/aconfmgr#215] [Q&A] [answered]
### General description of the problem:

I am trying to use the Whitelist method mentioned in the examples to backup some files from home. I tried to use the function `IgnorePathsExcept` like

```bash
white_list=(
    'pacman.conf'
)
IgnorePathsExcept '/etc' "${white_list[@]}"
```
I see that it works and only the `pacman.conf` file is backed up into files. This only works if I do not have any `IgnorePaths` command called on any of the parent directories in any of the previous files. 

Though, this does not work in my home directory. When I try

The `.bashrc` file is not backed up. The wiki says I can use it for the home directory, but that does not seem to work. Is this intended? It would be great to have a command to backup a specific file from a directory, just like the Whitelist, which can backup from any accessible directory across the system. Is the function `IgnotePathsExcept` supposed to work in that scenario? 

### Configuration:

```bash
01-helpers.sh

function IgnorePathsExcept() {
    # Ignore all path in given directory (first parameter) 
    # that do not math the given white list (second parameter)

    local search_dir=$1
    shift
    local white_list=("$@")
    local find_args=()
    local ignore_path
    for ignore_path in "${white_list[@]}"; do
        local base="$ignore_path"
        # Add all base paths to the argument list as well otherwise
        # -prune will prevent us from reaching the whitelisted files.
        while [ "$base" != '.' ]; do
            find_args+=(-path "$search_dir/$base" -o)
            base="$(dirname "$base")"
        done
    done
    # Find everything except given whitelist and the directory 
    # searched from.
    find "$search_dir" -not \( "${find_args[@]}" -path "$search_dir" \) -prune | \
    while read file; do 
        if [[ -d "$file" ]]; then
            IgnorePath "$file/*"
        else
            IgnorePath "$file"
        fi
    done
}
```

30-extra.sh
```
white_list=(
    'pacman.conf'
)
IgnorePathsExcept '/etc' "${white_list[@]}"

white_list=(
	'.bashrc'
)
IgnorePathsExcept '/home/saipavanchitta' "${white_list[@]}"
```

### Expected result:

`.bashrc` file is backed up from my home directory. 

### Actual result:

`.bashrc` file is not backed up.

### Log:

```bash
$aconfmgr save
: Collecting data...
:: Compiling user configuration...
::: Using configuration in /home/username/.config/aconfmgr
::: Sourcing /home/username/.config/aconfmgr/01-helpers.sh...
::: Sourcing /home/username/.config/aconfmgr/02-ignores.sh...
::: Sourcing /home/username/.config/aconfmgr/20-packages.sh...
::: Sourcing /home/username/.config/aconfmgr/30-home.sh...
::: Sourcing /home/saipavanchitta/.config/aconfmgr/99-unsorted.sh...
::: Done (248 native packages, 1 foreign packages, 1 files).
:: Inspecting system state...
::: Querying package list...
:::: Done.
::: Enumerating owned files...
:::: Done.
::: Searching for stray files...
:::: Done (3 stray files).
::: Cleaning up ignored files' directories...
:::: Done.
::: Searching for modified files...
:::: pacman: /etc/pacman.conf
:::: Done (1 modified files).
::: Reading file attributes...
:::: Reading file types...
:::: Reading file sizes...
:::: Reading file modes...
:::: Reading file owners...
:::: Reading file groups...
:::: Done.
::: Checking disk space...
:::: Done.
::: Processing found files...
:::: Done.
::: Done (248 native packages, 1 foreign packages, 1 files).
:: Examining files...
::: Loading data...
:::: Done.
::: Comparing file data...
::: Done (0 only in system, 0 changed, 0 only in config).
:: Examining file properties...
::: Loading data...
:::: Done.
::: Comparing file properties...
:::: Done.
::: Done (0 only in system, 0 changed, 0 only in config).
:: Done.
: Saving configuration...
:: Examining packages...
::: Done.
:: Registering files...
::: Done.
:: Registering file properties...
::: Comparing file properties...
:::: Done.
::: Done.
:: Done (configuration unchanged).
```

**Comment by CyberShadow** [comment]
`/home` is a default ignored path, perhaps it needs to be removed from `ignore_paths` first?

**Reply by saipavanc** [reply]
Yes, I just checked the source code and saw it in `ignore_paths`. Is there a way to remove a given directory from `ignore_paths`? Otherwise, is it possible to make another list like `backup_paths` with all the files or paths that need to be backed up? 

Here is what I am thinking. The broad `ignore_paths` is a list of directories that are ignored by default, or any paths added by `IgnorePaths`. Though, from a user's perspective it is easier to let them define some specific files which need to be backed up. So, maybe add a new function like `AddPaths`, which can be used in the following way:

```bash
AddPaths '~/*'
``` 
This would backup my home directory, and perhaps also simplify the implementation to the function `IgnorePathsExcept`. Also, any paths added with `AddPaths` will be backed up even if its parent directory is present in `ignore_paths`.  What do you think?

**Reply by CyberShadow** [reply, marked as answer]
> Is there a way to remove a given directory from `ignore_paths`?

Yes, just add [code to remove undesired elements](https://www.google.com/search?q=bash+remove+element+from+array) to your configuration, or redefine the array entirely.

> What do you think?

From the top of my head, unfortunately that seems to be quite far from how things work under the hood, so unlikely to be reasonably implementable in that form.

**Reply by saipavanc** [reply]
Great, redefinition worked! Thank you!

**Reply by Davi-S** [reply]
> > Is there a way to remove a given directory from `ignore_paths`?
> 
> Yes, just add [code to remove undesired elements](https://www.google.com/search?client=q=bash+remove+element+from+array) to your configuration, or redefine the array entirely.
> 
> > What do you think?
> 
> From the top of my head, unfortunately that seems to be quite far from how things work under the hood, so unlikely to be reasonably implementable in that form.

This link is broken. Can you provide a working one?

**Reply by CyberShadow** [reply]
I fixed the link.
